A valid configuration from Micronaut Launch (Kotlin with Java 16) fails to build

### Description

A clear and concise description of what the bug is.

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Create project from a valid configuration from Micronaut Launch
![image](https://user-images.githubusercontent.com/442813/117373781-9883b400-aed4-11eb-8488-94d8281fa71a.png)

`curl --location --request GET 'https://launch.micronaut.io/create/default/com.example.demo?lang=KOTLIN&build=GRADLE&test=JUNIT&javaVersion=JDK_16' --output demo.zip`
2. Extract the zip and run `./gradlew run`
3. Build fails with following error:

```bash
$ ./gradlew run
> Task :kaptKotlin FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':kaptKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.gradle.internal.KaptExecution
   > java.lang.reflect.InvocationTargetException (no error message)

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/7.0/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 3s
3 actionable tasks: 3 executed

```

The output with `--stacktrace` as follows:

```bash
$ ./gradlew run --stacktrace
> Task :kaptKotlin FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':kaptKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.gradle.internal.KaptExecution
   > java.lang.reflect.InvocationTargetException (no error message)

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':kaptKotlin'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:187)
 at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
...
        at org.gradle.workers.internal.DefaultWorkerExecutor$$Lambda$1104/0x00000000edd8d670.call(Unknown Source)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runExecution(DefaultConditionalExecutionQueue.java:214)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runBatch(DefaultConditionalExecutionQueue.java:164)
        at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.run(DefaultConditionalExecutionQueue.java:131)
        ... 3 more
Caused by: java.lang.IllegalAccessError: Class org/jetbrains/kotlin/kapt3/base/KaptContext(unnamed module 0x00000000F8006C10) can not access class com/sun/tools/javac/util/Context(module jdk.compiler) because module module jdk.compiler does not export package com/sun/tools/javac/util to module unnamed module 0x00000000F8006C10
        at org.jetbrains.kotlin.kapt3.base.KaptContext.<init>(KaptContext.kt:29)
        at org.jetbrains.kotlin.kapt3.base.Kapt.kapt(Kapt.kt:33)
        ... 34 more


* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/7.0/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1s
3 actionable tasks: 1 executed, 2 up-to-date
```

It's probably because of https://youtrack.jetbrains.com/issue/KT-45545 however the Launch shouldn't allow kotlin and java 16 combination then.

### Expected Behaviour

A valid (allowed) config from launch page is expected to work

### Actual Behaviour

it fails to build

### Environment Information

- **Operating System**: Ubuntu 20.04
- **Micronaut Version:** 2.5.1
- **JDK Version:** 
```bash
$ java -version
openjdk version "16.0.1" 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-16.0.1+9 (build 16.0.1+9)
Eclipse OpenJ9 VM AdoptOpenJDK-16.0.1+9 (build openj9-0.26.0, JRE 16 Linux amd64-64-Bit Compressed References 20210421_23 (JIT enabled, AOT enabled)
OpenJ9   - b4cc246d9
OMR      - 162e6f729
JCL      - cea22090ecf based on jdk-16.0.1+9)

```

### Example Application

- Create via the command:

`curl --location --request GET 'https://launch.micronaut.io/create/default/com.example.demo?lang=KOTLIN&build=GRADLE&test=JUNIT&javaVersion=JDK_16' --output demo.zip`

--- a/starter-core/src/main/java/io/micronaut/starter/feature/build/gradle/Gradle.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/build/gradle/Gradle.java
@@ -17,6 +17,7 @@ package io.micronaut.starter.feature.build.gradle;
 
 import io.micronaut.starter.application.ApplicationType;
 import io.micronaut.starter.application.generator.GeneratorContext;
+import io.micronaut.starter.build.BuildProperties;
 import io.micronaut.starter.build.gradle.GradleBuild;
 import io.micronaut.starter.build.gradle.GradleBuildCreator;
 import io.micronaut.starter.build.gradle.GradlePlugin;
@@ -28,6 +29,7 @@ import io.micronaut.starter.feature.build.gradle.templates.gradleProperties;
 import io.micronaut.starter.feature.build.gradle.templates.settingsGradle;
 import io.micronaut.starter.feature.database.JpaFeature;
 import io.micronaut.starter.options.BuildTool;
+import io.micronaut.starter.options.JdkVersion;
 import io.micronaut.starter.options.Options;
 import io.micronaut.starter.template.BinaryTemplate;
 import io.micronaut.starter.template.RockerTemplate;
@@ -67,6 +69,9 @@ public class Gradle implements BuildFeature {
             generatorContext.addBuildPlugin(GradlePlugin.builder().id("org.jetbrains.kotlin.plugin.allopen").lookupArtifactId("kotlin-allopen").build());
             if (generatorContext.getFeatures().isFeaturePresent(JpaFeature.class)) {
                 generatorContext.addBuildPlugin(GradlePlugin.builder().id("org.jetbrains.kotlin.plugin.jpa").lookupArtifactId("kotlin-noarg").build());
+            }
+            if (generatorContext.getJdkVersion().equals(JdkVersion.JDK_16)) {
+                generatorContext.getBuildProperties().put("org.gradle.jvmargs", "--illegal-access=permit");
             }
         }
         if (generatorContext.getFeatures().language().isGroovy() || generatorContext.getFeatures().testFramework().isSpock()) {

--- a/starter-core/src/main/java/io/micronaut/starter/feature/build/maven/Maven.java
+++ b/starter-core/src/main/java/io/micronaut/starter/feature/build/maven/Maven.java
@@ -24,12 +24,17 @@ import io.micronaut.starter.feature.build.BuildFeature;
 import io.micronaut.starter.feature.build.gitignore;
 import io.micronaut.starter.feature.build.maven.templates.pom;
 import io.micronaut.starter.options.BuildTool;
+import io.micronaut.starter.options.JdkVersion;
 import io.micronaut.starter.options.Options;
 import io.micronaut.starter.template.BinaryTemplate;
 import io.micronaut.starter.template.RockerTemplate;
+import io.micronaut.starter.template.Template;
 import io.micronaut.starter.template.URLTemplate;
 
 import javax.inject.Singleton;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
 import java.util.Set;
 
 @Singleton
@@ -67,6 +72,20 @@ public class Maven implements BuildFeature {
                 mavenBuild
         )));
         generatorContext.addTemplate("gitignore", new RockerTemplate(".gitignore", gitignore.template()));
+
+        if (generatorContext.getFeatures().language().isKotlin() && generatorContext.getJdkVersion().equals(JdkVersion.JDK_16)) {
+            generatorContext.addTemplate("mvnJvmConfig", new Template() {
+                @Override
+                public String getPath() {
+                    return ".mvn/jvm.config";
+                }
+
+                @Override
+                public void write(OutputStream outputStream) throws IOException {
+                    outputStream.write("--illegal-access=permit\n".getBytes(StandardCharsets.UTF_8));
+                }
+            });
+        }
     }
 
     @Override

--- a/starter-core/src/test/groovy/io/micronaut/starter/feature/build/gradle/GradleSpec.groovy
+++ b/starter-core/src/test/groovy/io/micronaut/starter/feature/build/gradle/GradleSpec.groovy
@@ -8,8 +8,10 @@ import io.micronaut.starter.feature.build.gradle.templates.gradleProperties
 import io.micronaut.starter.feature.build.gradle.templates.settingsGradle
 import io.micronaut.starter.fixture.CommandOutputFixture
 import io.micronaut.starter.options.BuildTool
+import io.micronaut.starter.options.JdkVersion
 import io.micronaut.starter.options.Language
 import io.micronaut.starter.options.Options
+import io.micronaut.starter.options.TestFramework
 import spock.lang.Issue
 import spock.lang.Unroll
 
@@ -62,6 +64,34 @@ class GradleSpec extends BeanContextSpec implements CommandOutputFixture {
         then:
         buildGradle
         buildGradle.contains("tasks")
+    }
+
+    @Unroll
+    @Issue('https://github.com/micronaut-projects/micronaut-starter/issues/774')
+    void 'Workaround for Kotlin projects with Gradle and JDK16'() {
+        when:
+        def output = generate(ApplicationType.DEFAULT, new Options(Language.KOTLIN, TestFramework.JUNIT, BuildTool.GRADLE, JdkVersion.JDK_16))
+        def properties = output["gradle.properties"]
+
+        then:
+        properties
+        properties.contains("org.gradle.jvmargs=--illegal-access=permit")
+
+        when:
+        output = generate(ApplicationType.DEFAULT, new Options(Language.KOTLIN, TestFramework.JUNIT, BuildTool.GRADLE, JdkVersion.JDK_11))
+        properties = output["gradle.properties"]
+
+        then:
+        properties
+        !properties.contains("org.gradle.jvmargs=--illegal-access=permit")
+
+        when:
+        output = generate(ApplicationType.DEFAULT, new Options(Language.JAVA, TestFramework.JUNIT, BuildTool.GRADLE, JdkVersion.JDK_16))
+        properties = output["gradle.properties"]
+
+        then:
+        properties
+        !properties.contains("org.gradle.jvmargs=--illegal-access=permit")
     }
 
 }

--- a/starter-core/src/test/groovy/io/micronaut/starter/feature/build/maven/MavenSpec.groovy
+++ b/starter-core/src/test/groovy/io/micronaut/starter/feature/build/maven/MavenSpec.groovy
@@ -8,15 +8,17 @@ import io.micronaut.starter.build.maven.MavenBuild
 import io.micronaut.starter.build.maven.MavenCombineAttribute
 import io.micronaut.starter.feature.Features
 import io.micronaut.starter.feature.build.maven.templates.pom
+import io.micronaut.starter.fixture.CommandOutputFixture
 import io.micronaut.starter.options.BuildTool
 import io.micronaut.starter.options.JdkVersion
 import io.micronaut.starter.options.Language
 import io.micronaut.starter.options.Options
 import io.micronaut.starter.options.TestFramework
 import io.micronaut.starter.util.VersionInfo
+import spock.lang.Issue
 import spock.lang.Unroll
 
-class MavenSpec extends BeanContextSpec {
+class MavenSpec extends BeanContextSpec implements CommandOutputFixture {
 
     void 'test use defaults from parent pom'() {
         given:
@@ -136,6 +138,32 @@ class MavenSpec extends BeanContextSpec {
         ApplicationType.DEFAULT     | "jetty-server"                | "jetty"
         ApplicationType.DEFAULT     | "netty-server"                | "netty"
         ApplicationType.DEFAULT     | "undertow-server"             | "undertow"
+    }
+
+    @Unroll
+    @Issue('https://github.com/micronaut-projects/micronaut-starter/issues/774')
+    void 'Workaround for Kotlin projects with Maven and JDK16'() {
+        when:
+        def output = generate(ApplicationType.DEFAULT, new Options(Language.KOTLIN, TestFramework.JUNIT, BuildTool.MAVEN, JdkVersion.JDK_16))
+        def jvmConfig = output[".mvn/jvm.config"]
+
+        then:
+        jvmConfig
+        jvmConfig.contains("--illegal-access=permit")
+
+        when:
+        output = generate(ApplicationType.DEFAULT, new Options(Language.KOTLIN, TestFramework.JUNIT, BuildTool.MAVEN, JdkVersion.JDK_11))
+        jvmConfig = output[".mvn/jvm.config"]
+
+        then:
+        !jvmConfig
+
+        when:
+        output = generate(ApplicationType.DEFAULT, new Options(Language.JAVA, TestFramework.JUNIT, BuildTool.MAVEN, JdkVersion.JDK_16))
+        jvmConfig = output[".mvn/jvm.config"]
+
+        then:
+        !jvmConfig
     }
 
 }

